aapcs64/aapcs64.rst

```diff
@@ -1023,6 +1158,604 @@ The A64 branch instructions are unable to reach every destination in the address
+return”, where “exceptional return” includes things like thread
+cancelation and C++ exception handling.
+
+The ZA lazy saving scheme
+-------------------------
+
+Overview
+^^^^^^^^
+
+**(Alpha)**
+
+SME provides a piece of storage called “ZA” that can be enabled and
+disabled using a processor state bit called “\ `PSTATE.ZA`_\ ”.  This storage
+is `SVL.B`_ × `SVL.B`_ bytes in size.  It can be accessed both
+horizontally and vertically, with each horizontal and vertical
+“slice” containing `SVL.B`_ bytes.
+
+The size of ZA can therefore vary between implementations of the Arm
+Architecture.  For a 256-bit SME implementation, ZA can hold the same
+amount of data as the 32 `Scalable Vector Registers`_.  For a 512-bit
+SME implementation, ZA can handle twice as much data as the vector registers,
+and so on.
+
+Suppose that a subroutine S1 with live data in ZA calls a subroutine S2 that
+has no knowledge of S1.  If the AAPCS64 defined ZA to be “call-preserved”
+(“callee-saved”), S1 would need to save and restore ZA around S1's own
+use of ZA, in case S1's caller also had live data in ZA.  If the AAPCS64
+defined ZA to be “call-clobbered” (“caller-saved”), S1 would need to
+save and restore ZA around the call to S2, in case S2 also used ZA.
+However, nested uses of ZA are expected to be rare, so these saves and
+restores would usually be wasted work.
+
+The AAPCS64 therefore defines a “lazy saving” scheme that often reduces
+the total number of saves and restores compared to the two approaches
+above.  Informally, the scheme allows “ZA is call-preserved” to become a
+dynamic rather than a static property: if S2 `complies with the lazy saving
+scheme`_, S1 can test after the call to S2 whether the call did in fact
+preserve ZA.  If the call did not preserve ZA, S1 is able to restore the
+old contents of ZA from a known buffer.
+
+The procedure is as follows:
+
+1. S1 `sets up a lazy save buffer`_ before calling S2.
+
+2. If the call to S2 `returns normally`_, the call is guaranteed to have
+   either `preserved ZA`_ or “\ `committed the lazy save`_\ ” (meaning that
+   it has saved ZA to the lazy save buffer).
+
+3. S1 then checks whether the call preserved ZA or not: `TPIDR2_EL0`_ is
+   nonnull if the call did preserve ZA; TPIDR2_EL0 is null if the call
+   committed the lazy save.
+
+   If TPIDR2_EL0 is null, S1 can restore the old contents of ZA
+   from the lazy save buffer that it set up in step (1).
+
+4. S1 `abandons the lazy save`_ when it no longer requires the contents
+   of ZA to be saved.
+
+See `SME pseudo-code examples`_ for a pseudo-code version of this procedure.
+
+Note that subroutines do not need to behave like S1 in the procedure
+above.  They could instead choose to save ZA before the call to S2
+and restore ZA after the call.
+
+ZA save buffers
+^^^^^^^^^^^^^^^
+
+**(Alpha)**
+
+A “\ _`ZA save buffer`\ “ is an area of memory of size W×16×\ `SVL.B`_
+bytes for some cardinal W.  The value of W is a property of the save
```

Comment:
The size of ZA is defined as `SVL.B x SVL.B bytes`.

What is the reason for defining the ZA save buffer as `W x 16 x SVL.B bytes` instead of:
```  W x SVL.B bytes```
where
``` num_za_save_slices <= W <= SVL.B```

Reply:
The current definition allows the save and restore code to be a loop that contains 16 loads and stores.  16 is special because it's maximum value for a 128-bit SME implementation (and also because LDR and STR have an immediate range of [0, 15]).